File: typescript/src/output/parquet/statistics.ts
/**
 * Column statistics — the min, the max and the NULL count of a column chunk.
 *
 * This is what lets a reader SKIP a whole row group: asked for `price > 500`,
 * it consults the chunk's max and moves on without decoding a byte. Cheap for
 * us (we already hold every value) and a large win for whoever queries the
 * file.
 *
 * The danger is the reverse of most features: WRONG statistics are worse than
 * none at all. A max that is too low makes a reader skip a group that really
 * did contain matching rows, and the query silently returns fewer results — no
 * error, no warning. So the comparisons here follow Parquet's declared sort
 * orders exactly rather than JavaScript's defaults:
 *
 *   - BYTE_ARRAY (strings) compare as UNSIGNED BYTES of their UTF-8 encoding.
 *     JavaScript's `<` compares UTF-16 code units, which agrees only for
 *     ASCII — "я" vs "ё" would come out wrong, and this project generates
 *     Russian data.
 *   - DOUBLE skips NaN entirely: NaN compares false against everything, so
 *     letting one in would poison the bound.
 *   - Everything else is numeric on its stored representation, which for
 *     DECIMAL means the unscaled integer and for DATE the day number.
 *
 * We write `min_value`/`max_value` (fields 5/6) and not the deprecated
 * `min`/`max` (1/2): the old pair had ambiguous signedness that readers
 * disagreed about, and writing a field readers may misread is the same trap as
 * writing a wrong bound.
 * Spec: docs/specs/2026-07-19-parquet-statistics-and-encodings.md
 */

import type { ColumnType } from '../column-type.js';

import type { TypedValue } from './convert.js';
import { halfBits } from './plain.js';

export interface ColumnStatistics {
  /** PLAIN-encoded single value; absent when the chunk has no non-NULL value. */
  readonly minValue?: Uint8Array;
  readonly maxValue?: Uint8Array;
  readonly nullCount: number;
}

const utf8 = new TextEncoder();

/** Unsigned byte-wise comparison — Parquet's sort order for BYTE_ARRAY. */
export function compareBytes(a: Uint8Array, b: Uint8Array): number {
  const shared = Math.min(a.length, b.length);
  for (let i = 0; i < shared; i++) {
    const x = a[i] ?? 0;
    const y = b[i] ?? 0;
    if (x !== y) return x < y ? -1 : 1;
  }
  return a.length === b.length ? 0 : a.length < b.length ? -1 : 1;
}

/** PLAIN encoding of ONE value, as statistics store it (no length prefix). */
function encodeOne(type: ColumnType, value: TypedValue): Uint8Array {
  switch (type.kind) {
    case 'bool': {
      return Uint8Array.from([value === true ? 1 : 0]);
    }
    case 'int32':
    case 'date':
    case 'uint8':
    case 'uint16':
    case 'uint32': {
      const out = new Uint8Array(4);
      new DataView(out.buffer).setInt32(0, Number(value), true);
      return out;
    }
    case 'float': {
      const out = new Uint8Array(4);
      new DataView(out.buffer).setFloat32(0, Number(value), true);
      return out;
    }
    case 'float16': {
      const out = new Uint8Array(2);
      new DataView(out.buffer).setUint16(0, halfBits(Number(value)), true);
      return out;
    }
    case 'int64':
    case 'timestamp':
    case 'decimal':
    case 'uint64': {
      const out = new Uint8Array(8);
      new DataView(out.buffer).setBigInt64(0, value as bigint, true);
      return out;
    }
    case 'double': {
      const out = new Uint8Array(8);
      new DataView(out.buffer).setFloat64(0, Number(value), true);
      return out;
    }
    case 'string':
    case 'enum':
    case 'json': {
      return utf8.encode(String(value));
    }
    case 'uuid': {
      return value instanceof Uint8Array ? value : new Uint8Array(0);
    }
  }
}

/**
 * Order two present values of this column type. Returns undefined when the
 * type has no usable order for a value (only NaN, today), so the caller can
 * leave it out of the bounds entirely.
 */
function compareValues(type: ColumnType, a: TypedValue, b: TypedValue): number | undefined {
  switch (type.kind) {
    case 'bool':
      return (a === true ? 1 : 0) - (b === true ? 1 : 0);
    case 'int32':
    case 'date':
    case 'float':
    case 'float16':
    case 'double': {
      const x = Number(a);
      const y = Number(b);
      if (Number.isNaN(x) || Number.isNaN(y)) return undefined;
      return x === y ? 0 : x < y ? -1 : 1;
    }
    case 'uint8':
    case 'uint16':
      // Small unsigned kinds keep their true value in the INT32 slot.
      return Number(a) - Number(b);
    case 'uint32':
    case 'uint64': {
      // Stored as wrapped signed bits, so they must be compared UNSIGNED —
      // otherwise 2^63 would look smaller than 1 and the bound would be wrong.
      const width = type.kind === 'uint32' ? 32n : 64n;
      const x = BigInt.asUintN(Number(width), BigInt(a as number | bigint));
      const y = BigInt.asUintN(Number(width), BigInt(b as number | bigint));
      return x === y ? 0 : x < y ? -1 : 1;
    }
    case 'int64':
    case 'timestamp':
    case 'decimal': {
      const x = a as bigint;
      const y = b as bigint;
      return x === y ? 0 : x < y ? -1 : 1;
    }
    case 'string':
    case 'enum':
    case 'json':
      return compareBytes(utf8.encode(String(a)), utf8.encode(String(b)));
    case 'uuid':
      return compareBytes(a as Uint8Array, b as Uint8Array);
  }
}

/** True when a value cannot take part in a bound (NaN only, for now). */
function unusable(type: ColumnType, value: TypedValue): boolean {
  const floaty = type.kind === 'double' || type.kind === 'float' || type.kind === 'float16';
  return floaty && Number.isNaN(Number(value));
}

/**
 * Min, max and NULL count for one column chunk.
 *
 * `present` holds the non-NULL values in row order; `nullCount` is supplied by
 * the caller because for a list column the NULLs live in the definition levels
 * rather than in the value array.
 */
export function computeStatistics(
  type: ColumnType,
  present: readonly TypedValue[],
  nullCount: number,
): ColumnStatistics {
  let min: TypedValue | undefined;
  let max: TypedValue | undefined;

  for (const value of present) {
    if (value === null || unusable(type, value)) continue;
    if (min === undefined || (compareValues(type, value, min) ?? 0) < 0) min = value;
    if (max === undefined || (compareValues(type, value, max) ?? 0) > 0) max = value;
  }

  if (min === undefined || max === undefined) return { nullCount };
  return {
    minValue: encodeOne(type, min),
    maxValue: encodeOne(type, max),
    nullCount,
  };
}

File: rust/src/output/parquet/statistics.rs
//! The min, the max and the NULL count of a column chunk.
//!
//! This is what lets a reader skip a whole row group: asked for `price > 500`, it
//! reads the chunk's maximum and moves on without decoding a byte. Cheap to
//! produce — every value is already in hand — and a large win for whoever queries
//! the file.
//!
//! The danger runs the other way from most features: wrong statistics are worse
//! than none. A maximum that is too low makes a reader skip a group that did
//! contain matching rows, and the query returns fewer results with no error and
//! no warning. So the comparisons here follow Parquet's declared sort orders
//! rather than the language's defaults — byte arrays compare as unsigned UTF-8
//! bytes, NaN never takes part in a bound, and the unsigned kinds are compared
//! unsigned even though they are stored in signed slots.
//!
//! Only `min_value`/`max_value` are written, never the deprecated `min`/`max`:
//! the old pair had ambiguous signedness that readers disagreed about, and
//! writing a field readers may misread is the same trap as writing a wrong bound.

use std::cmp::Ordering;

use super::convert::Value;
use super::plain;
use crate::output::column_type::{ColumnType, Kind};

/// PLAIN-encoded bounds; `None` when the chunk holds no non-NULL value at all.
pub struct Stats {
    pub min_value: Option<Vec<u8>>,
    pub max_value: Option<Vec<u8>>,
    pub null_count: i64,
}

/// Min, max and NULL count for one column chunk.
///
/// `null_count` is supplied by the caller because for a list column the NULLs
/// live in the definition levels rather than among the values.
pub fn compute(ty: &ColumnType, present: &[Option<Value>], null_count: i64) -> Stats {
    let mut min: Option<&Value> = None;
    let mut max: Option<&Value> = None;

    for value in present.iter().flatten() {
        if unusable(ty, value) {
            continue;
        }
        if min.is_none_or_greater(ty, value) {
            min = Some(value);
        }
        if max.is_none_or_less(ty, value) {
            max = Some(value);
        }
    }

    match (min, max) {
        (Some(min), Some(max)) => Stats {
            min_value: Some(encode_one(ty, min)),
            max_value: Some(encode_one(ty, max)),
            null_count,
        },
        _ => Stats {
            min_value: None,
            max_value: None,
            null_count,
        },
    }
}

/// Two small helpers so the loop above reads as the rule it implements.
trait Bound {
    fn is_none_or_greater(&self, ty: &ColumnType, candidate: &Value) -> bool;
    fn is_none_or_less(&self, ty: &ColumnType, candidate: &Value) -> bool;
}

impl Bound for Option<&Value> {
    fn is_none_or_greater(&self, ty: &ColumnType, candidate: &Value) -> bool {
        match self {
            None => true,
            Some(current) => compare(ty, candidate, current) == Ordering::Less,
        }
    }

    fn is_none_or_less(&self, ty: &ColumnType, candidate: &Value) -> bool {
        match self {
            None => true,
            Some(current) => compare(ty, candidate, current) == Ordering::Greater,
        }
    }
}

/// Unsigned byte-wise comparison — Parquet's sort order for a byte array.
pub fn compare_bytes(a: &[u8], b: &[u8]) -> Ordering {
    let shared = a.len().min(b.len());
    for i in 0..shared {
        if a[i] != b[i] {
            return a[i].cmp(&b[i]);
        }
    }
    a.len().cmp(&b.len())
}

/// PLAIN encoding of ONE value, as statistics store it — no length prefix.
fn encode_one(ty: &ColumnType, value: &Value) -> Vec<u8> {
    match (ty.kind, value) {
        (Kind::Bool, Value::Bool(v)) => vec![u8::from(*v)],
        (Kind::Int32 | Kind::Date | Kind::UInt8 | Kind::UInt16 | Kind::UInt32, Value::Int(v)) => {
            plain::int32(&[*v])
        }
        (Kind::Float, Value::Double(v)) => plain::floats(&[*v]),
        (Kind::Float16, Value::Double(v)) => plain::float16(&[*v]),
        (Kind::Int64 | Kind::Timestamp | Kind::Decimal | Kind::UInt64, Value::Long(v)) => {
            plain::int64(&[*v])
        }
        (Kind::Double, Value::Double(v)) => plain::doubles(&[*v]),
        (Kind::String | Kind::Enum | Kind::Json, Value::Text(v)) => v.as_bytes().to_vec(),
        (Kind::Uuid, Value::Bytes(v)) => v.clone(),
        _ => Vec::new(),
    }
}

/// Order two present values of this column type, following Parquet's rules for
/// it.
fn compare(ty: &ColumnType, a: &Value, b: &Value) -> Ordering {
    match (ty.kind, a, b) {
        (Kind::Bool, Value::Bool(x), Value::Bool(y)) => x.cmp(y),
        // The small unsigned kinds keep their true value in the signed slot.
        (Kind::Int32 | Kind::Date | Kind::UInt8 | Kind::UInt16, Value::Int(x), Value::Int(y)) => {
            x.cmp(y)
        }
        (Kind::Float | Kind::Float16 | Kind::Double, Value::Double(x), Value::Double(y)) => {
            if x.is_nan() || y.is_nan() {
                Ordering::Equal
            } else {
                x.partial_cmp(y).unwrap_or(Ordering::Equal)
            }
        }
        // Stored as wrapped signed bits, so compared unsigned — otherwise a value
        // above 2^31 would look smaller than one, and the bound would exclude
        // real rows.
        (Kind::UInt32, Value::Int(x), Value::Int(y)) => (*x as u32).cmp(&(*y as u32)),
        (Kind::UInt64, Value::Long(x), Value::Long(y)) => (*x as u64).cmp(&(*y as u64)),
        (Kind::Int64 | Kind::Timestamp | Kind::Decimal, Value::Long(x), Value::Long(y)) => x.cmp(y),
        (Kind::String | Kind::Enum | Kind::Json, Value::Text(x), Value::Text(y)) => {
            compare_bytes(x.as_bytes(), y.as_bytes())
        }
        (Kind::Uuid, Value::Bytes(x), Value::Bytes(y)) => compare_bytes(x, y),
        _ => Ordering::Equal,
    }
}

/// A value that cannot take part in a bound. NaN only, for now.
fn unusable(ty: &ColumnType, value: &Value) -> bool {
    matches!(ty.kind, Kind::Double | Kind::Float | Kind::Float16)
        && matches!(value, Value::Double(v) if v.is_nan())
}
